0000-frontend-development-infrastructure-and-workflow-changes.md

```diff
@@ -0,0 +1,73 @@
+
+# Detailed design
+[design]: #detailed-design
+
+Reactjs uses javascript to compose components and render HTML. Data is passed to the component via props, which are immutable. Component state is mutable and managed by the component. When state changes the component is rerendered.
+
+Note: for those familiar with Angular 1, React components are similar to Angular directives.
+
+The [Fixes #13424 - WIP c3 patternfly react implementation PR] (https://github.com/theforeman/foreman/pull/3603) provides a POC for integrating Reactjs with foreman.
+
+Top level components added to an HTML page (mounted) with the `ReactDOM` package `render` method. This is the point at which the Rails view passes the baton to Reactjs.
+
+The paradigm for using React components in Rails views implemented in the POC is as follows:
+
+>The `mount_react_component` helper in the `ReactjsHelper` module is used in the Rails view to which the component is to be added. The helper has 3 paramaters: a string identifier for the specific component to be mounted, a jquery selector that is used to find the DOM element into which to inject the component and, optionally, initial data from the controller in JSON format. See *'app/views/statistics/index.html.erb'* in the POC for an example.
+
+        <div id="statistics"></div>
+         
+        <%= mount_react_component('StatisticsChartsList', '#statistics', @data)%>
+
+> The helper injects inline javascript code into the Rails view which calls the `mount` method on the `reactMounter` object exposed on the global name space with the the parameters described above.
+
+> The `reactMounter` object manages a collection of outwardly facing components. The `mount` method finds the markup for the component specified by the identifier, adds the initial data (if provided) as props, finds the DOM element specified by the jquery selector and calls the `ReactDOM render` method with the component markup and a reference to the DOM element as parameters.
+
+> ReactDOM adds the component to the HTML page and Reactjs takes over from here.
+
+With this paradigm a contributor working in Ruby/Rails could use a previously developed component without great difficulty.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Remaining with the current infrastructure and code as implemented is not viable. 
+
+The options are to stick with the current infrastructure and rewrite current javascript code base or to integrate another library or framework.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Opinionated frameworks such as Angular 1, Angular 2 and Ember were considered. As the author of this RFC has particular knowledge and experience with that framework, Angular 1 was used as a proxy for the opinionated frameworks. The opinionated frameworks provide more built in features including application state management, but at the price of flexibility.
```

Comment:
@ehelms Flexibility in the case that no specific  comprehensive solution fits the requirements of the problem flexibility is the possibility of tailor a solution to the problem which is composed of several separate tools. (Can be risky in itself and would also require more skilled and experienced developers.)

Flexibility also pertains to how well the library/framework works with others and the extent to which the library/framework takes over the application.

No disrespect meant to angular. I've done 3 very successful greenfields projects in angular 1. None of those projects had server side rendering.

Foreman is going to be a server side rendered, not single page, Rails application  for a very long time if not forever. When I reason about using angular in this situation I can't come up with a paradigm that I think would be satisfactory.

Let's say I want to add a dynamic, interactive, client side list to an existing HTML page. With the React mounting paradigm I can compose an outwardly facing component in ES6, use the helper and mount that component. 

Let's try the same thing with angular. If I haven't already done so I will create a module and define a controller in the module which will manage a portion of the HTML page. That controller is associated with that portion of the page with a ng-controller directive. Then I will either have to write some angularized erb or use ng-include to insert a template. If I use a template I'm going to add template caching to the building and for testing. I  don't think that code would be very testable in any case.